{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "24181fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import time\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec56ec89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a user-agent header to identify your scraper\n",
    "user_agent = \"MyWebScraper/1.0 (+https://github.com/Adesuaayo/goodreads_webscraper)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2301cc04",
   "metadata": {},
   "outputs": [],
   "source": [
    "book_titles = []\n",
    "authors = []\n",
    "avg_ratings = []\n",
    "ratings = []\n",
    "published_years = []\n",
    "editions = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "710864c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error on page 7: ('Connection aborted.', TimeoutError(10060, 'A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond', None, 10060, None))\n",
      "Index error on page 29: list index out of range\n",
      "Index error on page 30: list index out of range\n",
      "Error on page 33: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=33&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "Error on page 50: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=50&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "Index error on page 99: list index out of range\n",
      "Error on page 102: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=102&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9A6D0850>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'))\n",
      "Error on page 112: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=112&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9A5F11C0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'))\n",
      "Error on page 119: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=119&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9A8C36D0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'))\n",
      "Error on page 139: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=139&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "Error on page 226: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "Error on page 301: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "Error on page 310: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=310&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "Error on page 460: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=460&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "Error on page 476: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=476&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D6A0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond'))\n",
      "Error on page 477: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=477&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D8E0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 478: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=478&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D340>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 479: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=479&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DCA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 480: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=480&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 481: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=481&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 482: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=482&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DD00>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 483: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=483&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DBE0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 484: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=484&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D6D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 485: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=485&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DD30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 486: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=486&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DAC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 487: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=487&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DC40>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 488: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=488&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA00>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 489: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=489&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DDF0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 490: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=490&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DDC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 491: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=491&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D7C0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 492: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=492&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D4F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 493: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=493&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA60>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 494: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=494&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D880>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 495: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=495&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D580>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 496: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=496&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DFA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 497: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=497&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DCD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 498: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=498&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D820>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 499: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=499&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D370>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 500: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=500&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DB80>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 501: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=501&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D880>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 502: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=502&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D940>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 503: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=503&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 504: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=504&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DCA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 505: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=505&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA60>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 506: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=506&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DBE0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 507: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=507&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D790>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 508: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=508&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DD00>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 509: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=509&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 510: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=510&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 511: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=511&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DD30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 512: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=512&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DB80>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 513: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=513&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D850>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 514: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=514&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D6D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 515: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=515&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DDF0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 516: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=516&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D880>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 517: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=517&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 518: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=518&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D550>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 519: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=519&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D790>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 520: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=520&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DCD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 521: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=521&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D340>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 522: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=522&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D940>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 523: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=523&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DAF0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 524: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=524&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DE50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 525: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=525&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DFA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 526: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=526&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D580>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 527: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=527&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA00>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 528: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=528&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DD90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 529: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=529&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 530: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=530&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D820>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 531: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=531&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA60>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 532: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=532&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D8E0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 533: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=533&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 534: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=534&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D4F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 535: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=535&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DE20>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 536: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=536&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DAC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 537: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=537&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D850>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 538: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=538&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D1C0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 539: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=539&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D580>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 540: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=540&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 541: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=541&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D550>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 542: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=542&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06D520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 543: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=543&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB9C06DA60>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 553: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=553&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1129)')))\n",
      "Error on page 777: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "Error on page 838: ('Connection aborted.', ConnectionResetError(10054, 'An existing connection was forcibly closed by the remote host', None, 10054, None))\n",
      "Index error on page 851: list index out of range\n",
      "Error on page 871: ('Connection aborted.', ConnectionAbortedError(10053, 'An established connection was aborted by the software in your host machine', None, 10053, None))\n",
      "Error on page 872: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=872&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 873: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=873&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 874: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=874&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4220>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 875: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=875&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4A30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 876: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=876&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40A0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 877: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=877&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 878: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=878&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4370>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 879: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=879&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4AC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 880: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=880&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 881: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=881&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4100>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 882: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=882&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4760>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 883: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=883&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4DC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 884: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=884&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4970>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 885: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=885&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4640>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 886: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=886&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 887: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=887&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 888: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=888&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 889: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=889&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB47F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 890: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=890&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E80>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 891: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=891&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB43D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 892: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=892&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4D30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 893: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=893&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4EB0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 894: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=894&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 895: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=895&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4670>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 896: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=896&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F40>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 897: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=897&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4B50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 898: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=898&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4940>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 899: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=899&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 900: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=900&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 901: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=901&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4220>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 902: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=902&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4A30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 903: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=903&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40A0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 904: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=904&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 905: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=905&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4370>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 906: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=906&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4AC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 907: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=907&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 908: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=908&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4100>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 909: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=909&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4760>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 910: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=910&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4DC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 911: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=911&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4970>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 912: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=912&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4640>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 913: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=913&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 914: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=914&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 915: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=915&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 916: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=916&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB47F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 917: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=917&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E80>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 918: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=918&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB43D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 919: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=919&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4D30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 920: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=920&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4EB0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 921: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=921&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 922: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=922&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4670>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 923: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=923&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F40>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 924: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=924&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4B50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 925: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=925&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4370>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 926: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=926&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4DC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 927: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=927&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 928: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=928&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 929: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=929&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4AC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 930: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=930&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4A30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 931: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=931&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4640>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 932: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=932&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4D30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 933: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=933&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 934: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=934&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4220>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 935: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=935&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4940>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 936: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=936&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4EB0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 937: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=937&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F10>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 938: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=938&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4100>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 939: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=939&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 940: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=940&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 941: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=941&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 942: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=942&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F70>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 943: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=943&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F40>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 944: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=944&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB43D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 945: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=945&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4670>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 946: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=946&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 947: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=947&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40A0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 948: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=948&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4970>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 949: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=949&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB47F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 950: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=950&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 951: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=951&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4B50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 952: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=952&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4370>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 953: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=953&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4DC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 954: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=954&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 955: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=955&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 956: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=956&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4AC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 957: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=957&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4A30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 958: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=958&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4640>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 959: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=959&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4D30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 960: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=960&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 961: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=961&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4220>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 962: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=962&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4940>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 963: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=963&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4EB0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 964: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=964&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F10>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 965: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=965&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4100>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 966: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=966&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 967: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=967&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 968: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=968&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 969: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=969&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F70>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 970: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=970&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F40>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 971: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=971&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB43D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 972: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=972&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4670>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 973: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=973&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 974: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=974&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB40A0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 975: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=975&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4970>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 976: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=976&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB47F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 977: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=977&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4490>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 978: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=978&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4B50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 979: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=979&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4370>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 980: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=980&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4DC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 981: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=981&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4520>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 982: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=982&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4E50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 983: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=983&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4AC0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 984: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=984&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4A30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 985: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=985&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4640>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 986: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=986&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4D30>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 987: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=987&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FA0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 988: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=988&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4220>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 989: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=989&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4940>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 990: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=990&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4EB0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 991: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=991&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F10>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 992: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=992&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4100>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 993: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=993&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 994: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=994&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 995: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=995&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 996: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=996&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F70>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 997: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=997&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4F40>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 998: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=998&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4220>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 999: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=999&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB49D0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n",
      "Error on page 1000: HTTPSConnectionPool(host='www.goodreads.com', port=443): Max retries exceeded with url: /search?page=1000&q=data+analysis&qid=LtFU5Ojc1o&tab=books (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001CB99DB4FD0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))\n"
     ]
    }
   ],
   "source": [
    "pages_to_scrape = 1000\n",
    "\n",
    "# Specify the delay between requests in seconds (e.g., 2 seconds)\n",
    "request_delay = 3\n",
    "\n",
    "# Loop through the pages to scrape\n",
    "for page in range(1, pages_to_scrape + 1):\n",
    "    \n",
    "    # Construct the URL for the current page\n",
    "    url = \"https://www.goodreads.com/search?page=\" + str(page) + \"&q=data+analysis&qid=LtFU5Ojc1o&tab=books\"\n",
    "   \n",
    "    try:\n",
    "        # Send an HTTP GET request to the URL with the user-agent header\n",
    "        headers = {\"User-Agent\": user_agent}\n",
    "        response = requests.get(url, headers=headers).text\n",
    "\n",
    "        # Parse the HTML content using BeautifulSoup\n",
    "        soup = BeautifulSoup(response, \"html5lib\")\n",
    "    \n",
    "        # Check for server errors or maintenance\n",
    "        if soup.title and \"service unavailable\" in soup.title.text.lower():\n",
    "            print(f\"Server error on page {page}. Skipping...\")\n",
    "            continue\n",
    "\n",
    "        # Select the table containing the list of books\n",
    "        table = soup.find_all(\"tbody\")[0]\n",
    "\n",
    "        # Loop through the rows of the table\n",
    "        for row in table.find_all(\"tr\"):\n",
    "            cells = row.find_all(\"td\")[1]\n",
    "\n",
    "            # Extract book title\n",
    "            title = cells.find(\"a\").find(\"span\").text\n",
    "            book_titles.append(title)\n",
    "\n",
    "            # Extract author's name\n",
    "            author = cells.find(\"a\", class_=\"authorName\").text\n",
    "            authors.append(author)\n",
    "            \n",
    "\n",
    "            #rating\n",
    "            all_ratings = cells.find_all('span', class_ = 'minirating')\n",
    "            all_ratings_text = all_ratings[0].text.strip()\n",
    "            pattern_2 = re.compile(r\"(\\d\\.?\\d*)\\savg\")\n",
    "            avg_ratings.append(pattern_2.search(all_ratings_text).group(1))\n",
    "\n",
    "            #n_ratings\n",
    "            pattern_4 = re.compile(r\"(\\d\\,?\\d*) rating\")\n",
    "            ratings_matches = pattern_4.search(all_ratings_text)\n",
    "            ratings.append(ratings_matches.group(1) if ratings_matches else 0)  \n",
    "\n",
    "\n",
    "#             # Extract average rating\n",
    "#             avg_rating = cells.find(\"span\", class_=\"greyText smallText uitext\").text.split()[0]\n",
    "#             avg_ratings.append(avg_rating)\n",
    "\n",
    "#             # Extract rating\n",
    "#             rating = cells.find(\"span\", class_=\"greyText smallText uitext\").text.split()[4]\n",
    "#             ratings.append(rating)\n",
    "\n",
    "            # Extract published year, handling cases where it may not be in the expected format\n",
    "            year_info = cells.find(\"span\", class_=\"greyText smallText uitext\").text.split()\n",
    "            year = None\n",
    "            for item in year_info:\n",
    "                if item.isdigit() and len(item) == 4:\n",
    "                    year = item\n",
    "                    break\n",
    "            if year:\n",
    "                published_years.append(year)\n",
    "            else:\n",
    "                published_years.append(0)  # Handle cases where year is not found\n",
    "\n",
    "            # Extract edition information\n",
    "            edition = cells.find(\"span\", class_=\"greyText smallText uitext\").text.split()[-2]\n",
    "            editions.append(edition)\n",
    "\n",
    "        # Sleep to add a delay between requests\n",
    "        time.sleep(request_delay)\n",
    "    \n",
    "    except requests.exceptions.RequestException as e:\n",
    "        # Handle HTTP request errors (e.g., connection issues)\n",
    "        print(f\"Error on page {page}: {e}\")\n",
    "\n",
    "    except IndexError as e:\n",
    "        # Handle \"list index out of range\" error\n",
    "        print(f\"Index error on page {page}: {e}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        # Handle other unexpected errors\n",
    "        print(f\"Unexpected error on page {page}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c007fabe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# After scraping all pages, we create a DataFrame from the collected data\n",
    "data = {\n",
    "    \"Title\": book_titles,\n",
    "    \"Author\": authors,\n",
    "    \"Average Rating\": avg_ratings,\n",
    "    \"Rating\": ratings,\n",
    "    \"Year Published\": published_years,\n",
    "    \"Edition\": editions\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1af705b6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Title</th>\n",
       "      <th>Author</th>\n",
       "      <th>Average Rating</th>\n",
       "      <th>Rating</th>\n",
       "      <th>Year Published</th>\n",
       "      <th>Edition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Python for Data Analysis</td>\n",
       "      <td>Wes McKinney</td>\n",
       "      <td>4.16</td>\n",
       "      <td>2,178</td>\n",
       "      <td>2011</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bayesian Data Analysis</td>\n",
       "      <td>Andrew Gelman</td>\n",
       "      <td>4.22</td>\n",
       "      <td>522</td>\n",
       "      <td>1995</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Data Analysis with Open Source Tools: A Hands-...</td>\n",
       "      <td>Philipp K. Janert</td>\n",
       "      <td>4.08</td>\n",
       "      <td>313</td>\n",
       "      <td>2010</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R Cookbook: Proven Recipes for Data Analysis, ...</td>\n",
       "      <td>Paul Teetor</td>\n",
       "      <td>4.05</td>\n",
       "      <td>439</td>\n",
       "      <td>2011</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Data Analysis Using Regression and Multilevel/...</td>\n",
       "      <td>Andrew Gelman</td>\n",
       "      <td>4.35</td>\n",
       "      <td>274</td>\n",
       "      <td>2006</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Title             Author  \\\n",
       "0                           Python for Data Analysis       Wes McKinney   \n",
       "1                             Bayesian Data Analysis      Andrew Gelman   \n",
       "2  Data Analysis with Open Source Tools: A Hands-...  Philipp K. Janert   \n",
       "3  R Cookbook: Proven Recipes for Data Analysis, ...        Paul Teetor   \n",
       "4  Data Analysis Using Regression and Multilevel/...      Andrew Gelman   \n",
       "\n",
       "  Average Rating Rating Year Published Edition  \n",
       "0           4.16  2,178           2011      38  \n",
       "1           4.22    522           1995      18  \n",
       "2           4.08    313           2010      10  \n",
       "3           4.05    439           2011       7  \n",
       "4           4.35    274           2006      12  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goodreads = pd.DataFrame(data)\n",
    "\n",
    "# Display the first five rows of the dataframe\n",
    "goodreads.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eb783175",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15678"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(goodreads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e11df4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "goodreads.to_csv(\"data_analysis_books.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d8b9dc9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
